File: Android/src/build-backups/kapt3-20260412-000124/stubs/debug/selfgemma/talk/domain/roleplay/usecase/PromptAssembler.java
package selfgemma.talk.domain.roleplay.usecase;

@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000\u00a2\u0001\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\b\n\u0002\b\u0003\u0018\u0000 92\u00020\u0001:\u00019B\u0011\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0004\b\u0004\u0010\u0005Jh\u0010\f\u001a\u00020\r2\u0006\u0010\u000e\u001a\u00020\u000f2\b\u0010\u0010\u001a\u0004\u0018\u00010\u00112\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00020\u00140\u00132\f\u0010\u0015\u001a\b\u0012\u0004\u0012\u00020\u00160\u00132\b\b\u0002\u0010\u0017\u001a\u00020\r2\b\b\u0002\u0010\u0018\u001a\u00020\r2\b\b\u0002\u0010\u0019\u001a\u00020\u001a2\n\b\u0002\u0010\u001b\u001a\u0004\u0018\u00010\u001c2\b\b\u0002\u0010\u001d\u001a\u00020\u001eJt\u0010\u001f\u001a\u00020 2\u0006\u0010\u000e\u001a\u00020\u000f2\b\u0010\u0010\u001a\u0004\u0018\u00010\u00112\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00020\u00140\u00132\f\u0010\u0015\u001a\b\u0012\u0004\u0012\u00020\u00160\u00132\b\b\u0002\u0010\u0017\u001a\u00020\r2\b\b\u0002\u0010\u0018\u001a\u00020\r2\b\b\u0002\u0010\u0019\u001a\u00020\u001a2\n\b\u0002\u0010!\u001a\u0004\u0018\u00010\r2\n\b\u0002\u0010\u001b\u001a\u0004\u0018\u00010\u001c2\b\b\u0002\u0010\u001d\u001a\u00020\u001eJh\u0010\u001f\u001a\u00020 2\u0006\u0010\"\u001a\u00020#2\u0006\u0010$\u001a\u00020%2\b\u0010\u0010\u001a\u0004\u0018\u00010\u00112\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00020\u00140\u00132\f\u0010\u0015\u001a\b\u0012\u0004\u0012\u00020\u00160\u00132\b\b\u0002\u0010\u0017\u001a\u00020\r2\n\b\u0002\u0010&\u001a\u0004\u0018\u00010\'2\n\b\u0002\u0010\u001b\u001a\u0004\u0018\u00010\u001c2\b\b\u0002\u0010\u001d\u001a\u00020\u001eJ\u001c\u0010(\u001a\b\u0012\u0004\u0012\u00020\u00160\u00132\f\u0010)\u001a\b\u0012\u0004\u0012\u00020\u00160\u0013H\u0002J\u0014\u0010*\u001a\u00020\r*\u00020+2\u0006\u0010\"\u001a\u00020#H\u0002JN\u0010,\u001a\u00020-2\u0006\u0010\"\u001a\u00020#2\b\u0010\u0010\u001a\u0004\u0018\u00010\u00112\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00020\u00140\u00132\f\u0010.\u001a\b\u0012\u0004\u0012\u00020\u00160\u00132\u0006\u0010\u0017\u001a\u00020\r2\u0006\u0010\u0018\u001a\u00020\r2\u0006\u0010/\u001a\u000200H\u0002J\u0018\u00101\u001a\u0004\u0018\u000102*\u0004\u0018\u0001032\u0006\u0010/\u001a\u000200H\u0002J\f\u00104\u001a\u00020\r*\u000202H\u0002J\f\u00104\u001a\u00020\r*\u000205H\u0002J\u0013\u00106\u001a\u00020\r*\u0004\u0018\u000107H\u0002\u00a2\u0006\u0002\u00108R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u000bX\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006:"}, d2 = {"Lselfgemma/talk/domain/roleplay/usecase/PromptAssembler;", "", "tokenEstimator", "Lselfgemma/talk/domain/roleplay/usecase/TokenEstimator;", "<init>", "(Lselfgemma/talk/domain/roleplay/usecase/TokenEstimator;)V", "characterBookRuntime", "Lselfgemma/talk/domain/roleplay/usecase/StCharacterBookRuntime;", "materialBuilder", "Lselfgemma/talk/domain/roleplay/usecase/PromptMaterialBuilder;", "contextBudgetPlanner", "Lselfgemma/talk/domain/roleplay/usecase/ContextBudgetPlanner;", "assemble", "", "role", "Lselfgemma/talk/domain/roleplay/model/RoleCard;", "summary", "Lselfgemma/talk/domain/roleplay/model/SessionSummary;", "memories", "", "Lselfgemma/talk/domain/roleplay/model/MemoryItem;", "recentMessages", "Lselfgemma/talk/domain/roleplay/model/Message;", "pendingUserInput", "generationTrigger", "userProfile", "Lselfgemma/talk/domain/roleplay/model/StUserProfile;", "contextProfile", "Lselfgemma/talk/domain/roleplay/model/ModelContextProfile;", "budgetMode", "Lselfgemma/talk/domain/roleplay/usecase/PromptBudgetMode;", "assembleForSession", "Lselfgemma/talk/domain/roleplay/usecase/PromptAssemblyResult;", "chatMetadataJson", "runtimeRole", "Lselfgemma/talk/domain/roleplay/model/StChatRuntimeRole;", "runtimeSession", "Lselfgemma/talk/domain/roleplay/model/StChatRuntimeSession;", "runtimeProfile", "Lselfgemma/talk/domain/roleplay/model/RoleRuntimeProfile;", "selectRecentMessages", "messages", "toSpeakerLabel", "Lselfgemma/talk/domain/roleplay/model/MessageSide;", "buildStScanContext", "Lselfgemma/talk/domain/roleplay/usecase/StWorldScanContext;", "dialogueWindow", "macroContext", "Lselfgemma/talk/domain/roleplay/usecase/StMacroContext;", "toDepthPrompt", "Lselfgemma/talk/domain/roleplay/usecase/DepthPromptInsertion;", "Lcom/google/gson/JsonObject;", "toPromptSection", "Lselfgemma/talk/domain/roleplay/usecase/StRuntimeDepthPromptInsertion;", "toPromptRoleName", "", "(Ljava/lang/Integer;)Ljava/lang/String;", "Companion", "app_debug"})
public final class PromptAssembler {
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.TokenEstimator tokenEstimator = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.StCharacterBookRuntime characterBookRuntime = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.PromptMaterialBuilder materialBuilder = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.ContextBudgetPlanner contextBudgetPlanner = null;
    @org.jetbrains.annotations.NotNull()
    public static final selfgemma.talk.domain.roleplay.usecase.PromptAssembler.Companion Companion = null;
    
    @javax.inject.Inject()
    public PromptAssembler(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.TokenEstimator tokenEstimator) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final java.lang.String assemble(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.RoleCard role, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.SessionSummary summary, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.MemoryItem> memories, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.Message> recentMessages, @org.jetbrains.annotations.NotNull()
    java.lang.String pendingUserInput, @org.jetbrains.annotations.NotNull()
    java.lang.String generationTrigger, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.StUserProfile userProfile, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.ModelContextProfile contextProfile, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.PromptBudgetMode budgetMode) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final selfgemma.talk.domain.roleplay.usecase.PromptAssemblyResult assembleForSession(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.RoleCard role, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.SessionSummary summary, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.MemoryItem> memories, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.Message> recentMessages, @org.jetbrains.annotations.NotNull()
    java.lang.String pendingUserInput, @org.jetbrains.annotations.NotNull()
    java.lang.String generationTrigger, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.StUserProfile userProfile, @org.jetbrains.annotations.Nullable()
    java.lang.String chatMetadataJson, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.ModelContextProfile contextProfile, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.PromptBudgetMode budgetMode) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final selfgemma.talk.domain.roleplay.usecase.PromptAssemblyResult assembleForSession(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.StChatRuntimeRole runtimeRole, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.StChatRuntimeSession runtimeSession, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.SessionSummary summary, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.MemoryItem> memories, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.Message> recentMessages, @org.jetbrains.annotations.NotNull()
    java.lang.String pendingUserInput, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.RoleRuntimeProfile runtimeProfile, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.ModelContextProfile contextProfile, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.PromptBudgetMode budgetMode) {
        return null;
    }
    
    private final java.util.List<selfgemma.talk.domain.roleplay.model.Message> selectRecentMessages(java.util.List<selfgemma.talk.domain.roleplay.model.Message> messages) {
        return null;
    }
    
    private final java.lang.String toSpeakerLabel(selfgemma.talk.domain.roleplay.model.MessageSide $this$toSpeakerLabel, selfgemma.talk.domain.roleplay.model.StChatRuntimeRole runtimeRole) {
        return null;
    }
    
    private final selfgemma.talk.domain.roleplay.usecase.StWorldScanContext buildStScanContext(selfgemma.talk.domain.roleplay.model.StChatRuntimeRole runtimeRole, selfgemma.talk.domain.roleplay.model.SessionSummary summary, java.util.List<selfgemma.talk.domain.roleplay.model.MemoryItem> memories, java.util.List<selfgemma.talk.domain.roleplay.model.Message> dialogueWindow, java.lang.String pendingUserInput, java.lang.String generationTrigger, selfgemma.talk.domain.roleplay.usecase.StMacroContext macroContext) {
        return null;
    }
    
    private final selfgemma.talk.domain.roleplay.usecase.DepthPromptInsertion toDepthPrompt(com.google.gson.JsonObject $this$toDepthPrompt, selfgemma.talk.domain.roleplay.usecase.StMacroContext macroContext) {
        return null;
    }
    
    private final java.lang.String toPromptSection(selfgemma.talk.domain.roleplay.usecase.DepthPromptInsertion $this$toPromptSection) {
        return null;
    }
    
    private final java.lang.String toPromptSection(selfgemma.talk.domain.roleplay.usecase.StRuntimeDepthPromptInsertion $this$toPromptSection) {
        return null;
    }
    
    private final java.lang.String toPromptRoleName(java.lang.Integer $this$toPromptRoleName) {
        return null;
    }
    
    @kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000\f\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0003\b\u0086\u0003\u0018\u00002\u00020\u0001B\t\b\u0002\u00a2\u0006\u0004\b\u0002\u0010\u0003\u00a8\u0006\u0004"}, d2 = {"Lselfgemma/talk/domain/roleplay/usecase/PromptAssembler$Companion;", "", "<init>", "()V", "app_debug"})
    public static final class Companion {
        
        private Companion() {
            super();
        }
    }
}

File: Android/src/app/src/main/java/selfgemma/talk/domain/roleplay/usecase/PromptAssembler.kt
package selfgemma.talk.domain.roleplay.usecase

import com.google.gson.JsonObject
import javax.inject.Inject
import selfgemma.talk.domain.roleplay.model.MemoryItem
import selfgemma.talk.domain.roleplay.model.Message
import selfgemma.talk.domain.roleplay.model.MessageKind
import selfgemma.talk.domain.roleplay.model.MessageSide
import selfgemma.talk.domain.roleplay.model.MessageStatus
import selfgemma.talk.domain.roleplay.model.ModelContextProfile
import selfgemma.talk.domain.roleplay.model.RoleCard
import selfgemma.talk.domain.roleplay.model.RoleRuntimeProfile
import selfgemma.talk.domain.roleplay.model.SessionSummary
import selfgemma.talk.domain.roleplay.model.StChatRuntimeRole
import selfgemma.talk.domain.roleplay.model.StChatRuntimeSession
import selfgemma.talk.domain.roleplay.model.StPersonaDescriptionPosition
import selfgemma.talk.domain.roleplay.model.StUserProfile
import selfgemma.talk.domain.roleplay.model.cardData
import selfgemma.talk.domain.roleplay.model.exampleDialoguesRaw
import selfgemma.talk.domain.roleplay.model.name
import selfgemma.talk.domain.roleplay.model.personaDescriptionForAuthorNote
import selfgemma.talk.domain.roleplay.model.personaDescriptionForDepthPrompt
import selfgemma.talk.domain.roleplay.model.personaDescription
import selfgemma.talk.domain.roleplay.model.summary
import selfgemma.talk.domain.roleplay.model.systemPrompt
import selfgemma.talk.domain.roleplay.model.tags
import selfgemma.talk.domain.roleplay.model.toStChatRuntimeRole
import selfgemma.talk.domain.roleplay.model.userPersonaDescription
import selfgemma.talk.domain.roleplay.model.worldSettings

private const val RECENT_DIALOGUE_TOKEN_BUDGET = 1800
private const val ST_DEFAULT_WORLD_INFO_SCAN_DEPTH = 4

class PromptAssembler @Inject constructor(private val tokenEstimator: TokenEstimator) {
  private val characterBookRuntime = StCharacterBookRuntime(tokenEstimator)
  private val materialBuilder = PromptMaterialBuilder(tokenEstimator)
  private val contextBudgetPlanner = ContextBudgetPlanner(tokenEstimator)

  fun assemble(
    role: RoleCard,
    summary: SessionSummary?,
    memories: List<MemoryItem>,
    recentMessages: List<Message>,
    pendingUserInput: String = "",
    generationTrigger: String = "normal",
    userProfile: StUserProfile = StUserProfile(),
    contextProfile: ModelContextProfile? = null,
    budgetMode: PromptBudgetMode = PromptBudgetMode.FULL,
  ): String {
    return assembleForSession(
      role = role,
      summary = summary,
      memories = memories,
      recentMessages = recentMessages,
      pendingUserInput = pendingUserInput,
      generationTrigger = generationTrigger,
      userProfile = userProfile,
      chatMetadataJson = null,
      contextProfile = contextProfile,
      budgetMode = budgetMode,
    ).prompt
  }

  fun assembleForSession(
    role: RoleCard,
    summary: SessionSummary?,
    memories: List<MemoryItem>,
    recentMessages: List<Message>,
    pendingUserInput: String = "",
    generationTrigger: String = "normal",
    userProfile: StUserProfile = StUserProfile(),
    chatMetadataJson: String? = null,
    contextProfile: ModelContextProfile? = null,
    budgetMode: PromptBudgetMode = PromptBudgetMode.FULL,
  ): PromptAssemblyResult {
    val runtimeRole = role.toStChatRuntimeRole(userProfile = userProfile)
    val runtimeSession =
      StChatRuntimeSession(
        chatMetadataJson = chatMetadataJson,
        generationTrigger = generationTrigger,
      )
    return assembleForSession(
      runtimeRole = runtimeRole,
      runtimeSession = runtimeSession,
      summary = summary,
      memories = memories,
      recentMessages = recentMessages,
      pendingUserInput = pendingUserInput,
      runtimeProfile = role.runtimeProfile,
      contextProfile = contextProfile,
      budgetMode = budgetMode,
    )
  }

  fun assembleForSession(
    runtimeRole: StChatRuntimeRole,
    runtimeSession: StChatRuntimeSession,
    summary: SessionSummary?,
    memories: List<MemoryItem>,
    recentMessages: List<Message>,
    pendingUserInput: String = "",
    runtimeProfile: RoleRuntimeProfile? = null,
    contextProfile: ModelContextProfile? = null,
    budgetMode: PromptBudgetMode = PromptBudgetMode.FULL,
  ): PromptAssemblyResult {
    val dialogueWindow = selectRecentMessages(recentMessages)
    val macroContext = runtimeRole.toStMacroContext()
    val scanContext =
      buildStScanContext(
        runtimeRole = runtimeRole,
        summary = summary,
        memories = memories,
        dialogueWindow = dialogueWindow,
        pendingUserInput = pendingUserInput,
        generationTrigger = runtimeSession.generationTrigger,
        macroContext = macroContext,
      )
    val cardData = runtimeRole.cardData()
    val resolvedCharacterBook =
      characterBookRuntime.resolve(
        book = cardData.character_book,
        context = scanContext,
        macroContext = macroContext,
        chatMetadataJson = runtimeSession.chatMetadataJson,
        chatLength = recentMessages.count { it.kind == MessageKind.TEXT && it.side != MessageSide.SYSTEM },
      )
    val coreDepthPrompt = cardData.extensions.toDepthPrompt(macroContext)
    val combinedExampleDialogue =
        buildList {
          addAll(resolvedCharacterBook.exampleBefore)
          addAll(runtimeRole.exampleDialoguesRaw().split("\n\n").map(String::trim).filter(String::isNotBlank).map(macroContext::substitute).filter { it.isNotBlank() })
          addAll(resolvedCharacterBook.exampleAfter)
        }
        .joinToString("\n")
        .trim()
    val postHistoryBlock =
      buildList {
          if (runtimeRole.userProfile.personaDescriptionPosition == StPersonaDescriptionPosition.TOP_AN) {
            runtimeRole.userProfile.personaDescriptionForAuthorNote()?.let(::add)
          }
          addAll(resolvedCharacterBook.authorNoteBefore)
          cardData.post_history_instructions
            ?.let(macroContext::substitute)
            ?.trim()
            ?.takeIf(String::isNotBlank)
            ?.let(::add)
          addAll(resolvedCharacterBook.authorNoteAfter)
          if (runtimeRole.userProfile.personaDescriptionPosition == StPersonaDescriptionPosition.BOTTOM_AN) {
            runtimeRole.userProfile.personaDescriptionForAuthorNote()?.let(::add)
          }
        }
        .joinToString("\n")
        .trim()
    val depthPromptBlock =
      buildList {
          coreDepthPrompt?.toPromptSection()?.let(::add)
          runtimeRole.userProfile.personaDescriptionForDepthPrompt()
            ?.let(macroContext::substitute)
            ?.takeIf(String::isNotBlank)
            ?.let { personaDepthPrompt ->
              add(
                buildString {
                  appendLine("role=${runtimeRole.userProfile.personaDescriptionRole.toPromptRoleName()} depth=${runtimeRole.userProfile.personaDescriptionDepth}")
                  append(personaDepthPrompt)
                }
              )
            }
          addAll(resolvedCharacterBook.depthPrompts.map { it.toPromptSection() })
        }
        .joinToString("\n\n")
        .trim()

    val material =
      materialBuilder.build(
        runtimeRole = runtimeRole,
        runtimeProfile = runtimeProfile,
        summary = summary,
        memories = memories,
        recentMessages = recentMessages,
        macroContext = macroContext,
        resolvedCharacterBook = resolvedCharacterBook,
        postHistoryBlock = postHistoryBlock,
        depthPromptBlock = depthPromptBlock,
        combinedExampleDialogue = combinedExampleDialogue,
      )
    val plan =
      contextBudgetPlanner.plan(
        material = material,
        contextProfile = contextProfile,
        preferredMode = budgetMode,
      )
    return PromptAssemblyResult(
      prompt = plan.prompt,
      updatedChatMetadataJson = resolvedCharacterBook.updatedChatMetadataJson,
      budgetReport = plan.report,
    )
  }

  private fun selectRecentMessages(messages: List<Message>): List<Message> {
    val filtered =
      messages.filter { message ->
        message.kind == MessageKind.TEXT &&
          message.side != MessageSide.SYSTEM &&
          message.status != MessageStatus.FAILED &&
          message.content.isNotBlank()
      }

    if (filtered.isEmpty()) {
      return emptyList()
    }

    val selected = mutableListOf<Message>()
    var tokenCount = 0

    for (message in filtered.asReversed()) {
      val nextTokenCount =
        tokenCount + tokenEstimator.estimate(message.content) + tokenEstimator.estimate(message.side.name)
      if (selected.isNotEmpty() && nextTokenCount > RECENT_DIALOGUE_TOKEN_BUDGET) {
        break
      }

      selected += message
      tokenCount = nextTokenCount
    }

    return selected.asReversed()
  }

  private fun MessageSide.toSpeakerLabel(runtimeRole: StChatRuntimeRole): String {
    return when (this) {
      MessageSide.USER -> runtimeRole.userName
      MessageSide.ASSISTANT -> runtimeRole.name()
      MessageSide.SYSTEM -> "System"
    }
  }

  private fun buildStScanContext(
    runtimeRole: StChatRuntimeRole,
    summary: SessionSummary?,
    memories: List<MemoryItem>,
    dialogueWindow: List<Message>,
    pendingUserInput: String,
    generationTrigger: String,
    macroContext: StMacroContext,
  ): StWorldScanContext {
    val core = runtimeRole.card
    val data = runtimeRole.cardData()
    val recentMessagesNewestFirst =
      buildList {
        pendingUserInput.trim().takeIf(String::isNotBlank)?.let(::add)
        dialogueWindow
          .asReversed()
          .mapTo(this) { message ->
            "${message.side.toSpeakerLabel(runtimeRole)}: ${message.content.trim()}"
          }
      }

    return StWorldScanContext(
      roleName = runtimeRole.name(),
      roleTags = runtimeRole.tags(),
      generationTrigger = generationTrigger,
      recentMessagesNewestFirst = recentMessagesNewestFirst,
      userPersonaDescription = macroContext.substitute(runtimeRole.userPersonaDescription()),
      characterDescription = macroContext.substitute(runtimeRole.summary()),
      characterPersonality =
        macroContext.substitute(
          data.personality.orEmpty().ifBlank { core.personality.orEmpty().ifBlank { runtimeRole.personaDescription() } }
        ),
      characterDepthPrompt = data.extensions.toDepthPrompt(macroContext)?.prompt.orEmpty(),
      scenario =
        macroContext.substitute(
          data.scenario.orEmpty().ifBlank { core.scenario.orEmpty().ifBlank { runtimeRole.worldSettings() } }
        ),
      creatorNotes = macroContext.substitute(data.creator_notes.orEmpty().ifBlank { core.creatorcomment.orEmpty() }),
      sessionSummary = summary?.summaryText.orEmpty(),
      memories = memories.map { it.content.trim() }.filter(String::isNotBlank),
    )
  }

  private fun JsonObject?.toDepthPrompt(macroContext: StMacroContext): DepthPromptInsertion? {
    val depthPrompt = this?.getAsJsonObject("depth_prompt") ?: return null
    val prompt =
      macroContext.substitute(depthPrompt.get("prompt")?.takeIf { it.isJsonPrimitive }?.asString).trim()
    if (prompt.isBlank()) {
      return null
    }
    return DepthPromptInsertion(
      prompt = prompt,
      depth = depthPrompt.get("depth")?.takeIf { it.isJsonPrimitive }?.asInt ?: ST_DEFAULT_WORLD_INFO_SCAN_DEPTH,
      role =
        depthPrompt.get("role")?.takeIf { it.isJsonPrimitive }?.asString?.trim()?.ifBlank { null }
          ?: "system",
    )
  }

  private fun DepthPromptInsertion.toPromptSection(): String {
    return buildString {
      appendLine("role=$role depth=$depth")
      append(prompt)
    }
  }

  private fun StRuntimeDepthPromptInsertion.toPromptSection(): String {
    return buildString {
      appendLine("role=$role depth=$depth")
      append(prompts.joinToString("\n"))
    }
  }

  private fun Int?.toPromptRoleName(): String {
    return when (this) {
      1 -> "user"
      2 -> "assistant"
      else -> "system"
    }
  }

  companion object {
  }
}

private data class DepthPromptInsertion(
  val prompt: String,
  val depth: Int,
  val role: String,
)
